teps/0010-optional-workspaces.md

```diff
@@ -0,0 +1,367 @@
+result in an empty directory being placed in its absence, resulting in an
+incorrectly passing behaviour gate.
+
+### Example: `git-clone` Catalog Task
+
+The `git-clone` Catalog Task already supports the `creds-init` approach to
+credential handling. Here is an example updating the Task to include support
+for an explicit `.ssh` directory to be provided as an optional Workspace:
+
+```yaml
+# A number of fields which have no bearing on the behaviour of this example
+# have been removed from the original Catalog entry.
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: git-clone
+spec:
+  workspaces:
+    - name: output
+      description: The git repo will be cloned onto the volume backing this workspace
+    - name: ssh-directory
+      description: An .ssh directory with private key and known_hosts file.
+      optional: true
+  params:
+    - name: url
+      description: git url to clone
+      type: string
+    - name: revision
+      description: git revision to checkout (branch, tag, sha, ref…)
+      type: string
+      default: master
+    - name: subdirectory
+      description: subdirectory inside the "output" workspace to clone the git repo into
+      type: string
+      default: ""
+  results:
+    - name: commit
+      description: The precise commit SHA that was fetched by this Task
+  steps:
+    - name: clone
+      image: gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/git-init:v0.14.2
+      script: |
+        CHECKOUT_DIR="$(workspaces.output.path)/$(params.subdirectory)"
+
+        # Here we only use the credentials provided via the ssh-directory Workspace
+        # if creds-init did not initialize any SSH creds already.
+        if [ ! -d "$HOME/.ssh" ] ; then
+          if [ "$(workspaces.ssh-directory.bound)" == "true" ]; then
+            cp -R "$(workspaces.ssh-directory.path)" "$HOME/.ssh"
+          fi
+        fi
+
+        /ko-app/git-init \
+          -url "$(params.url)" \
+          -revision "$(params.revision)" \
+          -path "$CHECKOUT_DIR"
+        cd "$CHECKOUT_DIR"
+        RESULT_SHA="$(git rev-parse HEAD | tr -d '\n')"
+        EXIT_CODE="$?"
+        if [ "$EXIT_CODE" != 0 ]
+        then
+          exit $EXIT_CODE
+        fi
+        # Make sure we don't add a trailing newline to the result!
+        echo -n "$RESULT_SHA" > $(results.commit.path)
+
+```
+
+## Test Plan
+
+Unit tests for the behaviour of a Workspace being bound and being omitted
+when the declaration is optional.
+
+Unit tests for the variable interpolation code when a Workspace is omitted.
+
+Unit tests for the variable interpolation of a Workspace's bound state.
+
+Unit tests for the behaviour of an omitted Workspace when there is a
+Default Workspace configured.
+
+E2E tests can take the form of YAML examples that demonstrate the
+syntax for declaring a Workspace optional, providing and omitting a Workspace,
+as well as usage of optional Workspace variables.
+
+## Drawbacks
+
+TBD
```

Comment:
a couple things that come to mind: (cant quite decide if these are risks or drawbacks)

-  a task might run extra containers that it doesnt need in the case of an optional workspace, e.g. imagine the last step in a task copies test results onto an optional workspace, and when the workspace isn't provided, the step still runs but does nothing
- might be a bit more confusing to understand what a Task actually requires? (only a bit tho probably)

Reply:
>     * a task might run extra containers that it doesnt need in the case of an optional workspace, e.g. imagine the last step in a task copies test results onto an optional workspace, and when the workspace isn't provided, the step still runs but does nothing

Is it a drawback though ? It doesn't change anything that could happen now (without this feature). You can have a step that does nothing but still runs (based on some params).

>     * might be a bit more confusing to understand what a Task actually requires? (only a bit tho probably)

We know what is optional for the task (optional workspace, params with default vaule), so what the task requires is whatever is not optional, nor has defaults. A UI could display that I guess.

Reply:
For the second point I agree completely this design means the workspace's `description` field has to pull double-duty: 1) It has to describe what the workspace is for  2) It has to describe how the Task behaviour changes if it's omitted.

Reply:
I've added this as a Drawback